{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "85dfd832",
   "metadata": {},
   "source": [
    "# Coletor de Informações GenBank"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04c7b8f9",
   "metadata": {},
   "source": [
    "###### by Pedro Souza"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bafcdd4",
   "metadata": {},
   "source": [
    "> <div align=\"justify\"> Coletar informações no GenBank visando criar um repositório um pouco mais robusto para as suas análises moleculares as vezes é complicado devido aos diversos parâmetros existentes para cada deposito e o jeito do GenBank apresentar essas informações. Visando melhorar a visualização inicial desse material pelos pesquisadores desenvolvi esse organizador de informações. Para utilizar basta baixar todo o conteudo que deseja triar utilizando a função compiladora do GenBank e salvar o arquivo completo (.gb ou Genbank), depois basta salvar uma cópia desse material em txt e utilizar a função  <b><font color='green'>GenBank_coletor()</font>.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc675f3c",
   "metadata": {},
   "source": [
    "## Extenções necessárias"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "aac38f4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50cacc13",
   "metadata": {},
   "source": [
    "## Funções necessárias"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b0829926",
   "metadata": {},
   "outputs": [],
   "source": [
    "def data_reader(file): # Função que realizará a leitura do arquivo Genbank\n",
    "    with open(file) as f:\n",
    "        GB = f.read()\n",
    "    \n",
    "    GB = re.sub('://', '', GB)\n",
    "    GB_split = GB.split('//')\n",
    "    \n",
    "    print(len(GB_split), 'possiveis sequências foram reconhecidas. \\n\\n' )\n",
    "    return GB_split\n",
    "\n",
    "\n",
    "def info_organizer(data): # Função que realizará a organização dos dados retirados do arquivo base\n",
    "    \n",
    "    prod = 0\n",
    "    a = 0\n",
    "    n = 0\n",
    "    j = []\n",
    "    t = []\n",
    "    seq = []\n",
    "    tam = []\n",
    "    time = []\n",
    "    sp = []\n",
    "    gene = []\n",
    "    title = []\n",
    "    journal = []\n",
    "    journal_temp = []\n",
    "    pais = []\n",
    "    \n",
    "    for y in data:\n",
    "        \n",
    "        if len(y) > 10: # Essa condicional remove qualquer erro na leitura dos registros, já que remove registros com menos de 10 caracteres\n",
    "            \n",
    "            lines = re.split('\\n+',y) # Para dar menos trabalho da análise dos registros divido os mesmos em linhas\n",
    "            \n",
    "            ng = len(re.findall('/product=',y)) # Alguns registros possuem mais de um gene utilizo esse comando para contar\n",
    "            ps = len(re.findall('/country',y))\n",
    "\n",
    "            for i in lines: # Com esse loop o programa vai procurar linha por linha as informações requisitadas\n",
    "                n += 1\n",
    "                \n",
    "                if re.search('LOCUS',i): # Essa parte coleta o código do GenBank, o tamanho e a data de depósito\n",
    "                    info = re.split('\\s+',i)\n",
    "                    seq.append(info[1])\n",
    "                    tam.append(info[2])\n",
    "                    time.append(info[7])\n",
    "                \n",
    "                \n",
    "                if re.search('SOURCE',i): # Essa parte coleta o orgânismos análisado\n",
    "                    info = re.split('\\s+',i)\n",
    "                    sp.append(info[1] + ' '+ info[2])\n",
    "                    #print(sp)\n",
    "                \n",
    "                if re.search('TITLE',i): # Essa parte colete a titulo do trabalho\n",
    "                    t.append(n)\n",
    "\n",
    "                if re.search('JOURNAL',i): # Essa parte coleta informação sobre a revista, caso o trabalho já tenha sido aceito\n",
    "                    journal_temp.append(i)\n",
    "                    j.append(n)\n",
    "                \n",
    "                \n",
    "            if ng == 1: # Aqui entramos na seleçõo dos genes\n",
    "                for i in lines:\n",
    "                    if re.findall('/product=',i):\n",
    "                        if re.search('small subunit ribosomal RNA',i) or re.search('18S ribosomal RNA',i) or re.search('small subunit ribosomal',i):\n",
    "                            gene.append('18S')\n",
    "                        elif re.search('large subunit ribosomal RNA',i):\n",
    "                            gene.append('28S')\n",
    "                        elif re.search('cytochrome c oxidase subunit 1',i) or re.search('cytochrome oxidase subunit 1',i):\n",
    "                            gene.append('COI-1')\n",
    "                        else:\n",
    "                            gene_temp = i\n",
    "                            gene_temp = re.sub('\\s+',' ', gene_temp)\n",
    "                            gene_temp = re.sub('/product=','', gene_temp)\n",
    "                            gene_temp = re.sub('\"','', gene_temp)\n",
    "                            \n",
    "                            gene.append(gene_temp)\n",
    "            elif ng < 1:\n",
    "                if re.search('/note=',y):\n",
    "                    gene.append('Possível fragmento')\n",
    "            else:\n",
    "                #print(a, ': Multi-gene')\n",
    "                gene.append('Multi-gene')\n",
    "                    #gene.append(2)\n",
    "                    \n",
    "            if ps == 1:\n",
    "                for i in lines:\n",
    "                    if re.search('/country', i):\n",
    "                        pais_temp = i\n",
    "                        pais_temp = re.sub('\\s+',' ', pais_temp)\n",
    "                        pais_temp = re.sub('/country=','', pais_temp)\n",
    "                        pais_temp = re.sub('\"','', pais_temp)\n",
    "                        pais.append(pais_temp)\n",
    "                        \n",
    "            if ps < 1:\n",
    "                pais.append('SR')\n",
    "            \n",
    "            if ps > 1:\n",
    "                pais.append('Mais que um pais')\n",
    "\n",
    "            #print(a,t)\n",
    "\n",
    "            a += 1\n",
    "            prod = 0\n",
    "           \n",
    "            # A coleta de informações sobre os títulos e sobre as revistas apresentaram alguns problemas, como repetições ou estar incompleto.\n",
    "            \n",
    "            # Resolvendo titulo\n",
    "\n",
    "            h = int(t[0] - 1)\n",
    "            g = int(j[0] - 1)\n",
    "            titulo_temp = ''.join(lines[h:g])\n",
    "            titulo_temp = re.sub('TITLE', '', titulo_temp)\n",
    "            titulo_temp = re.sub('\\s+', ' ', titulo_temp)\n",
    "            title.append(titulo_temp)\n",
    "\n",
    "            n = 0\n",
    "            t = []\n",
    "            j = []\n",
    "\n",
    "            # Resolvendo Revista\n",
    "\n",
    "            j_temp = re.sub('JOURNAL','',journal_temp[0])\n",
    "            j_temp = re.sub('\\s+',' ',j_temp)\n",
    "\n",
    "            journal.append(j_temp)\n",
    "\n",
    "            journal_temp = []\n",
    "\n",
    "            #\n",
    "    \n",
    "    df = pd.DataFrame(list(zip(seq, sp, gene,tam, time,pais,title,journal)),\n",
    "               columns =['Sequência','Espécie','Gene', 'Tamanho', 'Data','País','Título','Revista'])\n",
    "    \n",
    "    print('Informações de {} sequências foram recuperadas.'.format(len(df)))\n",
    "    print('Destas, {}, são possíveis sequências fragmentadas, {}, apresentam mais de um gene e {}, apresentam algum tipo de erro de identificação do gene.'.format(len(df.query('Gene == \"Possível fragmento\"')), len(df.query('Gene == \"Multi-gene\"')), len(df.query('Gene == \"Error\"'))))\n",
    "    \n",
    "    return(df)\n",
    "\n",
    "def GenBank_coletor(path): # Função final que compila as duas primeiras e salva o resultado em um arquivo excel e retorna, tambem, os dados para análise posterior\n",
    "    GB = data_reader(path)\n",
    "    df = info_organizer(GB)\n",
    "    df.to_excel('GenBank_data_out.xlsx')\n",
    "    print('\\n\\n' + 'Seus dados foram organizados no arquivo \"GenBank_data_out.xlsx\".')\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01063782",
   "metadata": {},
   "source": [
    "> Para utilizar o organizador basta passar para a função o arquivo que deseja ser análisado, e.g.:\n",
    "    <b><p style=\"color:blue\"> GenBank_coletor('dataset_Peritrichia_out2022.txt') </p></b>\n",
    "    \n",
    "> A função retornará um dataframe organizado que será salvo na sua pasta e também pode ser adicionada a um objeto. Além disso, retornará diversas informações sobre o sucesso do processo, como número de sequências indentificadas, análisadas e etc.\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d1972be6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1344 possiveis sequências foram reconhecidas. \n",
      "\n",
      "\n",
      "Informações de 1343 sequências foram recuperadas.\n",
      "Destas, 134, são possíveis sequências fragmentadas, 354, apresentam mais de um gene e 0, apresentam algum tipo de erro de identificação do gene.\n",
      "\n",
      "\n",
      "Seus dados foram organizados no arquivo \"GenBank_data_out.xlsx\".\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequência</th>\n",
       "      <th>Espécie</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Tamanho</th>\n",
       "      <th>Data</th>\n",
       "      <th>País</th>\n",
       "      <th>Título</th>\n",
       "      <th>Revista</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OP289656</td>\n",
       "      <td>Epistylis pygmaeum</td>\n",
       "      <td>18S</td>\n",
       "      <td>1586</td>\n",
       "      <td>14-OCT-2022</td>\n",
       "      <td>China: Jining</td>\n",
       "      <td>Integrative Studies on Three Epibiotic Episty...</td>\n",
       "      <td>Protist 173 (5), 125909 (2022)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OP289655</td>\n",
       "      <td>Epistylis daphniae</td>\n",
       "      <td>18S</td>\n",
       "      <td>1648</td>\n",
       "      <td>14-OCT-2022</td>\n",
       "      <td>China: Jining</td>\n",
       "      <td>Integrative Studies on Three Epibiotic Episty...</td>\n",
       "      <td>Protist 173 (5), 125909 (2022)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OP289654</td>\n",
       "      <td>Epistylis weishanensis</td>\n",
       "      <td>18S</td>\n",
       "      <td>1647</td>\n",
       "      <td>14-OCT-2022</td>\n",
       "      <td>China: Jining</td>\n",
       "      <td>Integrative Studies on Three Epibiotic Episty...</td>\n",
       "      <td>Protist 173 (5), 125909 (2022)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ON158510</td>\n",
       "      <td>Carchesium cyclopidarum</td>\n",
       "      <td>18S</td>\n",
       "      <td>1544</td>\n",
       "      <td>26-SEP-2022</td>\n",
       "      <td>China: Jining</td>\n",
       "      <td>The taxonomy and molecular phylogeny of two e...</td>\n",
       "      <td>Eur J Protistol 86, 125921 (2022)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ON158509</td>\n",
       "      <td>Epistylis qingdaoensis</td>\n",
       "      <td>18S</td>\n",
       "      <td>1613</td>\n",
       "      <td>26-SEP-2022</td>\n",
       "      <td>China: Qingdao</td>\n",
       "      <td>The taxonomy and molecular phylogeny of two e...</td>\n",
       "      <td>Eur J Protistol 86, 125921 (2022)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1338</th>\n",
       "      <td>AF335515</td>\n",
       "      <td>Epistylis wenrichi</td>\n",
       "      <td>18S</td>\n",
       "      <td>1723</td>\n",
       "      <td>16-NOV-2001</td>\n",
       "      <td>SR</td>\n",
       "      <td>Phylogenetic relationships of the subclass pe...</td>\n",
       "      <td>J. Eukaryot. Microbiol. 48 (5), 583-587 (2001)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1339</th>\n",
       "      <td>AF335514</td>\n",
       "      <td>Epistylis chrysemydis</td>\n",
       "      <td>18S</td>\n",
       "      <td>1727</td>\n",
       "      <td>16-NOV-2001</td>\n",
       "      <td>SR</td>\n",
       "      <td>Phylogenetic relationships of the subclass pe...</td>\n",
       "      <td>J. Eukaryot. Microbiol. 48 (5), 583-587 (2001)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1340</th>\n",
       "      <td>AF335513</td>\n",
       "      <td>Epistylis hentscheli</td>\n",
       "      <td>18S</td>\n",
       "      <td>1729</td>\n",
       "      <td>16-NOV-2001</td>\n",
       "      <td>SR</td>\n",
       "      <td>Phylogenetic relationships of the subclass pe...</td>\n",
       "      <td>J. Eukaryot. Microbiol. 48 (5), 583-587 (2001)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1341</th>\n",
       "      <td>AF070701</td>\n",
       "      <td>Vorticella microstoma</td>\n",
       "      <td>18S</td>\n",
       "      <td>1688</td>\n",
       "      <td>01-JUL-1998</td>\n",
       "      <td>SR</td>\n",
       "      <td>Phylogenetic relationships of Vorticella conv...</td>\n",
       "      <td>Unpublished</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1342</th>\n",
       "      <td>AF070700</td>\n",
       "      <td>Vorticella convallaria</td>\n",
       "      <td>18S</td>\n",
       "      <td>1698</td>\n",
       "      <td>01-JUL-1998</td>\n",
       "      <td>SR</td>\n",
       "      <td>Phylogenetic relationships of Vorticella conv...</td>\n",
       "      <td>Unpublished</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1343 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sequência                  Espécie Gene Tamanho         Data  \\\n",
       "0     OP289656       Epistylis pygmaeum  18S    1586  14-OCT-2022   \n",
       "1     OP289655       Epistylis daphniae  18S    1648  14-OCT-2022   \n",
       "2     OP289654   Epistylis weishanensis  18S    1647  14-OCT-2022   \n",
       "3     ON158510  Carchesium cyclopidarum  18S    1544  26-SEP-2022   \n",
       "4     ON158509   Epistylis qingdaoensis  18S    1613  26-SEP-2022   \n",
       "...        ...                      ...  ...     ...          ...   \n",
       "1338  AF335515       Epistylis wenrichi  18S    1723  16-NOV-2001   \n",
       "1339  AF335514    Epistylis chrysemydis  18S    1727  16-NOV-2001   \n",
       "1340  AF335513     Epistylis hentscheli  18S    1729  16-NOV-2001   \n",
       "1341  AF070701    Vorticella microstoma  18S    1688  01-JUL-1998   \n",
       "1342  AF070700   Vorticella convallaria  18S    1698  01-JUL-1998   \n",
       "\n",
       "                 País                                             Título  \\\n",
       "0       China: Jining   Integrative Studies on Three Epibiotic Episty...   \n",
       "1       China: Jining   Integrative Studies on Three Epibiotic Episty...   \n",
       "2       China: Jining   Integrative Studies on Three Epibiotic Episty...   \n",
       "3       China: Jining   The taxonomy and molecular phylogeny of two e...   \n",
       "4      China: Qingdao   The taxonomy and molecular phylogeny of two e...   \n",
       "...               ...                                                ...   \n",
       "1338               SR   Phylogenetic relationships of the subclass pe...   \n",
       "1339               SR   Phylogenetic relationships of the subclass pe...   \n",
       "1340               SR   Phylogenetic relationships of the subclass pe...   \n",
       "1341               SR   Phylogenetic relationships of Vorticella conv...   \n",
       "1342               SR   Phylogenetic relationships of Vorticella conv...   \n",
       "\n",
       "                                              Revista  \n",
       "0                      Protist 173 (5), 125909 (2022)  \n",
       "1                      Protist 173 (5), 125909 (2022)  \n",
       "2                      Protist 173 (5), 125909 (2022)  \n",
       "3                   Eur J Protistol 86, 125921 (2022)  \n",
       "4                   Eur J Protistol 86, 125921 (2022)  \n",
       "...                                               ...  \n",
       "1338   J. Eukaryot. Microbiol. 48 (5), 583-587 (2001)  \n",
       "1339   J. Eukaryot. Microbiol. 48 (5), 583-587 (2001)  \n",
       "1340   J. Eukaryot. Microbiol. 48 (5), 583-587 (2001)  \n",
       "1341                                      Unpublished  \n",
       "1342                                      Unpublished  \n",
       "\n",
       "[1343 rows x 8 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GenBank_coletor('dataset_Peritrichia_out2022.txt')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
